### ISSUE TITLE
Confused Matrix Name

### ISSUE THREAD

**Issue by dumganhar** [cocos2d/cocos2d-x#16731]
Reported at http://discuss.cocos2d-x.org/t/cc-mvmatrix-help/20974

> Figured it out. Turns out CC_MVMatrix is just the model matrix & CC_PMatrix is the view-projection matrix. Those names are pretty mis-leading!

I think the developer in the forum is correct. We have used wrong names for `model(world), view, projection` matrix.

**Comment by dumganhar** [comment]
@ricardoquesada , any thoughts about this issue?

**Comment by ricardoquesada** [comment]
@dumganhar any suggestion? what names should we use? thanks.

**Comment by dumganhar** [comment]
@ricardoquesada , 
Currently, we have missed use the name for a long time in cocos2d-x 3.x. There're serval places refer to `Matrix`.

In OpenGL 1.x API, there are some stack operation APIs (deprecated in OpenGLES 2.0) called:

https://www.opengl.org/sdk/docs/man2/xhtml/glMatrixMode.xml
- void glMatrixMode(GLenum mode);
  
  The mode supports:
  
  _GL_MODELVIEW_
  Applies subsequent matrix operations to the modelview matrix stack.
  
  _GL_PROJECTION_
  Applies subsequent matrix operations to the projection matrix stack.
  
  _GL_TEXTURE_
  Applies subsequent matrix operations to the texture matrix stack.
  
  _GL_COLOR_
  Applies subsequent matrix operations to the color matrix stack.
- void glLoadMatrixf(const GLfloat *m);
- void glPushMatrix(void);
- void glPopMatrix(void);

The mode `GL_MODELVIEW` means to combine `View( or Camera) Matrix` with `Model (or World) Matrix` , its result `View * Model` transform will convert a point to camera coordinate.
And the mode `GL_PROJECTION` is only used for projection for converting a point in camera coordinate to device view port.

But in Cocos2d-x 3.x, we use
- `Node::_modelViewTransform`: It only used for `Model/World` transform, personally I like the name `_worldTransform` or `_worldMatrix`. GamePlay3d's node names it in this way (https://github.com/gameplay3d/GamePlay/blob/master/gameplay/src/Node.h#L750)
- `TrianglesCommand::_mv`, `MeshCommand::_mv`, `BatchCommand::_mv`, all places where there are `_mv`/`mv`/`modelViewTransform` means **only** _World Transform_.
- Builtin shaders: 
  
  GLProgram::_flags.usesP (CC_PMatrix in shader): Means whether to use `Projection` transform, but currently it stands for `View + Projection` transform.
  GLProgram::_flags.usesMV (CC_MVMatrix in shader): Means whether to use World(Model) + View transform, but currently it only stands for `World` transform.
  GLProgram::_flags.usesMVP (CC_MVPMatrix in shader): Means whether to use Projection \* View \* World transform, this is correct.

IMO, name is important for developer to understand our code. It should be clear enough.
But now we have made them stand for another thing. 

@ricardoquesada , current names are all ok to me, but they should be used as they are. For example:

The name `modelMatrix / worldMatrix`:  Only used for `Model -> World` transform
The name `viewMatrix`: Only used for `World -> View` transform
The name `modelViewMatrix`: It's eventually used for `Model -> World -> View` transform
The name `projectionMatrix`: Only used for `Projection` transform
The name `modelViewProjectMatrix`: It's eventually used for `Model -> World -> View -> Projection` transform

**Comment by ricardoquesada** [comment]
makes sense. I like. My only concern are the shaders. If we change it, it might break user-created shaders... is there a way we can do it in a backward compatible way?
thanks!

**Comment by dumganhar** [comment]
> If we change it, it might break user-created shaders... is there a way we can do it in a backward compatible way?

Yes, it will definitely break user's shader. For now, if we wanna correct these things, we need a good documentation for these changes.

**Comment by DavidDeSimone** [comment]
GLSL has #define, you could always pull a `#define CC_PMatrix viewProjectionMatrix` etc.

**Comment by ricardoquesada** [comment]
@DavidDeSimone thanks. the `#define` seems like a good compromise

**Comment by dumganhar** [comment]
@DavidDeSimone , it seems an approach.
@ricardoquesada , we could probably try it like what @DavidDeSimone  said.

**Comment by ricardoquesada** [comment]
@dumganhar do you want to create the PR? or do you want me to do it? thanks.

**Comment by dumganhar** [comment]
@ricardoquesada , yes, if you have time, please create a PR.
I have some other higher priority tasks to do. :)
